lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -551,6 +551,16 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
     };
 
     if (!isConfused) {
+      if (Name.getBaseName().userFacingName() == "Self") {
+        DeclName selfName(Context.getIdentifier("self"));
+        auto selfInstance = lookupUnqualified(DC, selfName, Loc, lookupOptions);
+        if (!selfInstance.empty()) {
+          ValueDecl *D = selfInstance.front().getValueDecl();
+          Expr *E = new (Context) DeclRefExpr(D, nameLoc, /*Implicit=*/false);
+          return new (Context) DynamicTypeExpr(Loc, Loc, E, Loc, Type());
```

Comment:
If you're inside a static method, Self should have the same type as self, however this gives it an extra level of metatype.

Also I think it would be better to produce a TypeExpr here and not a DynamicTypeExpr.

Reply:
Is there an easy way to detect when you're inside a static method? 
Doesn't it have to be a DynamicTypeExpr as the type is only known at run time (for classes)

Reply:
Yes, there is an isStatic() method on FuncDecl. However, if you use a TypeExpr and set its type to a DynamicSelfType you shouldn't need to check.

Reply:
Thanks. I'll take a look tomorrow.

Reply:
I've got Self as a type in the signature of value types working and fixed the double Metatype-ing of Self in a static function.  Tried to move to using a TypeExpr but couldn't get it to work. I've left what I tried in, perhaps someone has some pointers. Stdlib compiles and tests run as before.